### PR TITLE
[Keymap] Add dlip Layout to Chouchou keyboard

### PR DESCRIPTION
**Pull request by dlip**
## Description

Add dlip Layout to Chouchou keyboard

## Types of Changes

- [ ] Core
- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project: [**C**](https://docs.qmk.fm/#/coding_conventions_c), [**Python**](https://docs.qmk.fm/#/coding_conventions_python)
- [x] I have read the [**PR Checklist** document](https://docs.qmk.fm/#/pr_checklist) and have made the appropriate changes.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [ ] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

File: users/dlip/custom_keycodes.h [hunk 1]
```diff
@@ -0,0 +1,57 @@
+#pragma once
```
Comment (drashna) [inline users/dlip/custom_keycodes.h:1]: Could you add a GPL2+ compatible license header here?  All of the source files need to have a license explicitly listed.

For instance:
```c
// Copyright 2021 Your Name (@yourgithub)
// SPDX-License-Identifier: GPL-2.0-or-later
```

File: users/dlip/gamepad.h
Comment (drashna) [inline users/dlip/gamepad.h:1]: this should be a c file

**Review by drashna — CHANGES_REQUESTED** [review]
The keymap folder name and the userspace folder name should match here.  

Additionally, you're include a c file using an header file extension (`.h`), rather than the correct naming. 

Additionally, this really isn't the correct use for the userspace, and the code should be moved into the keymap.

**Comment by dlip** [comment]
@drashna regarding the incorrect use of the user space, the Taipo library is a custom combo engine meant to be able to be used across many keyboards so I think it doesn't make sense to put it just in the keymap. Is there a better option you can suggest?

**Comment by drashna** [comment]
If they're all in the same vendor folder, then `keyboards/(vendor)/lib` (or `common`) is what is recommended. 

As for using a userspace, at best it's really pushing what the userspace should be used for.   
However, it should, at worst, use the same name as the keymap folder. when these match, the userspace path is included automatically, and rules.mk and config.h are automatically processed.  Worst case, `USER_NAME := (name)` can be used to specify a different keymap name, but should be used only when necessary. 

Also, the .h files should be renamed to .c files, and a rules.mk file should use `SRC += gamepad.c` or the like.

The docs have a lot of good information and examples of how the userspace should be used, as well: 
https://docs.qmk.fm/#/feature_userspace

For instance:
```make
ifeq ($(strip $(JOYSTICK_ENABLE)), yes)
  SRC += gamepad.c
endif
```

**Comment by dlip** [comment]
Thanks @drashna I think I understand the user dir better now. I've updated the code.

**Comment by dlip** [comment]
@drashna could I gently poke you to have a look and my changes when you get the chance? :pray:

**Comment by Glodigit** [comment]
@dlip, I'm trying to implement code from this Taipo PR into my Lets Split configuration but it seems to be conflicting with something. I'm currently testing with one half at the moment, and the bottom 2 rows don't seem to register keypresses in my _BASE layer and my _TAIPO toggle key outputs "e" when trying to untoggle. Additionally, my LEDs are experiencing unexpected / delayed behaviour. My latest commit can be seen [here](https://github.com/Glodigit/vial-qmk/commit/4bc7d38a528c8e9dad53acd49a72c276186749d8). Could I ask for you to have a look?

On a seperate note, would it be possible to modify the code so that both the left and right halves can be chorded together and QMK sends out the left output then right output? For example, pressing "t" on the left half and "o" on the right simultaneously results in "to" being sent to the host. I haven't been able to try it, but it sounds like the implementation would output either "to" or "ot", depending on which hand finished the chord a handful of milliseconds earlier. This feature would allow for both an alternating-hands and steno-like typing style.

**Comment by dlip** [comment]
@Glodigit I can't see anything obviously wrong with your config, maybe its not playing nice with vial or something? I'd try to remove everything but taipo, and if it works add one thing back at a time to see whats causing it.

Regarding the steno type input, I can't think of a good way to do it without risking typing 'to' when you actually want to type 'ot'. As you learn the layout and get faster you'll start roll from side to side naturally and it would be frustrating to be 'corrected' like this. So I think its probably better to leave its current implementation, which is the side that releases one of the keys first wins.

**Comment by Glodigit** [comment]
Understandable. I hadn't considered the event where the fingers are still down on the right half at the time when the left chord starts.

**Pull request closed without merging**